clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/service/LambdaService.java

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright 2021 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.lambda.service;
+
+import com.amazonaws.auth.policy.Policy;
+import com.amazonaws.auth.policy.Statement;
+import com.amazonaws.services.lambda.AWSLambda;
+import com.amazonaws.services.lambda.model.*;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.netflix.spinnaker.clouddriver.aws.AmazonCloudProvider;
+import com.netflix.spinnaker.clouddriver.aws.data.ArnUtils;
+import com.netflix.spinnaker.clouddriver.aws.security.AmazonClientProvider;
+import com.netflix.spinnaker.clouddriver.aws.security.NetflixAmazonCredentials;
+import com.netflix.spinnaker.clouddriver.core.limits.ServiceLimitConfiguration;
+import com.netflix.spinnaker.clouddriver.lambda.service.config.LambdaServiceConfig;
+import com.netflix.spinnaker.kork.exceptions.SpinnakerException;
+import groovy.util.logging.Slf4j;
+import java.time.Clock;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+@Slf4j
+public class LambdaService {
+
+  private final AmazonClientProvider amazonClientProvider;
+  private final NetflixAmazonCredentials account;
+  private final String region;
+  private final int TIMEOUT_MINUTES;
+  private final int RETRIES;
+  private final Clock clock = Clock.systemDefaultZone();
+  private final ObjectMapper mapper;
+  private final ExecutorService executorService;
+
+  public LambdaService(
+      AmazonClientProvider amazonClientProvider,
+      NetflixAmazonCredentials account,
+      String region,
+      ObjectMapper mapper,
+      LambdaServiceConfig lambdaServiceConfig,
+      ServiceLimitConfiguration serviceLimitConfiguration) {
+    this.amazonClientProvider = amazonClientProvider;
+    this.account = account;
+    this.region = region;
+    this.mapper = mapper;
+    this.TIMEOUT_MINUTES = lambdaServiceConfig.getRetry().getTimeout();
+    this.RETRIES = lambdaServiceConfig.getRetry().getRetries();
+    this.executorService =
+        Executors.newFixedThreadPool(
+            computeThreads(serviceLimitConfiguration, lambdaServiceConfig));
+  }
```

Comment:
great to see this happen. thanks for the implementation. 

I feel like we can actually split it to 2 classes for better readability. One being the `LambdaService` class that includes all the calls to the API endpoints, and another as the `LambdaDataSupplier` or something, that has `getAllFunctions()` as well as all the functions to add `attributes` and doing `retries`, `computeThreads` etc.

Basically one specifically to interface with the API and one to do the heavy lifting of reading the data and filling the cache.

Reply:
I like this idea -- we can take this point away and implement further refactoring soon. I think at the moment, we want to try and get this in ASAP